linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -4495,6 +4492,27 @@ private boolean advanceRight(TInner right, TKey rightKey) {
     }
   }
 
+  public static int compareNullsLastForMergeJoin(@Nullable Comparable v0, @Nullable Comparable v1) {
+    return compareNullsLastForMergeJoin(v0, v1, null);
+  }
+
+  public static int compareNullsLastForMergeJoin(@Nullable Comparable v0, @Nullable Comparable v1,
+      @Nullable Comparator comparator) {
+    // Special code for mergeJoin algorithm: in case of two null values, they must not be
+    // considered as equal (otherwise the join would return incorrect results); instead, consider
+    // the first (left) value as "bigger", to advance on the right value and continue with the
+    // algorithm
+    if (v0 == null && v1 == null) {
+      return 1;
```

Comment:
I admit it is a hacky way to use comparators, and indeed it breaks its semantic, but I was not able to find a better way to fix this situation.

> Since we also use the same comparator to compare records from same side, it may be worth to also test the case that multi null key records from one side?
I'm not sure I understand... could you please provide an example?

Reply:
That is not correct, EnumerbleMergeJoin accepts inputs that are **already sorted** (by some other operator on its left/right child, typically an EnumerableSort, or some kind of InxedScan). So, the "first stage" that you mention does not happen inside EnumerbleMergeJoin, which only performs your "second stage". Therefore this "special comparator" (flagged as "MergeJoinCompartor" in the auxiliary methods that generate it) will only be used within EnumerbleMergeJoin code, and never outside of it.

Reply:
Thanks for the explanation, it sounds good to me. Maybe my impression is from that I searched the usage of this comparator, and found it's used here: https://github.com/apache/calcite/blob/cbf857e319a695fe96b1f3f5f38a779821c92cb3/linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java#L4393-L4401, and here: https://github.com/apache/calcite/blob/cbf857e319a695fe96b1f3f5f38a779821c92cb3/linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java#L4426-L4434

I guess you are more familiar than me with this part, if these two places are safe regarding this change, then I have no more concerns.

Reply:
Indeed, very well spotted @libenchao ! I had totally missed these comparisons applied on two keys from the same side. In fact, the MergeJoin patch is incorrect, and I can get an IllegalStateException in those places under certain conditions.
I will need to think about a different approach for the MergeJoin...
Thank you very much for the helpful review!

Reply:
@libenchao  I have committed a new approach that should handle this issue:
- Use a (control flow) BothValuesAreNullException in the Comparator to avoid breaking its semantics
- Use an EqualityComparer (i.e. nulls are considered equal) to compare keys from the same side

wdyt?

Reply:
The idea sounds good to me. I'm not sure what's happened with Github, there are a lot of unrelated commits identified by Github, now it's hard to view the changes, what do you think about rebase it to latest main?

Reply:
Done